### PR TITLE
[TSE] Implement the Market Making Strategy

### LINKED ISSUE #19: [TSE] Implement the Market Making Strategy
### Implement Market Making Strategy

**Description:**

This task involves implementing the Market Making Strategy, which should follow the interface defined in:

- `packages/tse-backend/src/modules/trading-strategy/strategy.interface.ts`

For reference, you can check the implementation in this file:

- [strategy.service.ts](https://github.com/Hu-Fi/Mr.Market/blob/main/server/src/modules/strategy/strategy.service.ts)

The minimum supported properties are outlined in:

- [strategy.dto.ts](https://github.com/Hu-Fi/Mr.Market/blob/main/server/src/modules/strategy/strategy.dto.ts#L52C1-L52C43)

### To Do

- [ ] Implement a method to create the Market Making strategy for the given user.
- [ ] Implement a method to stop the Market Making strategy for the given user.
- [ ] Implement a method to delete the Market Making strategy for the given user.

### Deliverables

- [ ] All tasks completed
- [ ] Unit tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by MWBlocky**
## Description

This Pull Request introduces the implementation of the **Market Making Strategy** for the TSE (Trading Strategy Execution) module. The Market Making Strategy is designed to provide liquidity by placing both buy and sell orders around the current market price. This implementation follows the interface defined in `strategy.interface.ts` and integrates with the existing system components.

## Summary of Changes

- **MarketMakingStrategy** class implementing the `Strategy` interface.
- **Create Market Making Strategy**: Added method to create entity with dto params.
- **Start Market Making Strategy**: Added a method to initialize and run the Market Making Strategy for a given user, including periodic placement of buy and sell orders based on the defined spread and order amount.
- **Pause Market Making Strategy**: Added functionality to pause the strategy for a given user.
- **Stop Market Making Strategy**: Implemented a method to stop the ongoing Market Making Strategy, including the cancellation of active orders.
- **Delete Market Making Strategy**: Added method to soft delete entity.
- **Order Placement Logic**: Incorporated logic to calculate the appropriate bid and ask prices based on the mid-price, with adjustments for the defined spread and layer configuration.
- **Unit Tests**: Created unit tests to verify the correct functionality of the strategy.

## Related Issues

Fixes #19

### CODE REVIEW

File: packages/tse-backend/src/common/entities/market-making.entity.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  id: number;
+
+  @Column()
+  userId: string;
+
+  @Column()
+  clientId: string;
+
+  @Column()
+  pair: string;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/entities/market-making.entity.ts:26]: I would split that probably to sideA and sideB or smt like that

File: packages/tse-backend/src/common/entities/market-making.entity.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  askSpread: number;
+
+  @Column('decimal', { precision: 10, scale: 2 })
+  orderAmount: number;
+
+  @Column()
+  checkIntervalSeconds: number;
+
+  @Column()
+  numberOfLayers: number;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/entities/market-making.entity.ts:44]: Check ehat's that

File: packages/tse-backend/src/common/utils/trading-strategy.utils.ts [hunk 3]
```diff
@@ -62,3 +68,122 @@ export const isExchangeSupported = (
...
+
+type PriceSourceFunction = (orderBook: any, ticker: any) => number;
+
+const priceSourceFunctions: Record<PriceSourceType, PriceSourceFunction> = {
+  [PriceSourceType.MID_PRICE]: (orderBook) =>
+    (orderBook.bids[0][0] + orderBook.asks[0][0]) / 2,
+  [PriceSourceType.BEST_ASK]: (orderBook) => orderBook.asks[0][0],
+  [PriceSourceType.BEST_BID]: (orderBook) => orderBook.bids[0][0],
+  [PriceSourceType.LAST_PRICE]: (_, ticker) => ticker.last,
+};
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/utils/trading-strategy.utils.ts:80]: What's that?

Comment (piotrswierzy) [reply packages/tse-backend/src/common/utils/trading-strategy.utils.ts:80]: Maybe add a comment

File: packages/tse-backend/src/common/utils/trading-strategy.utils.ts [hunk 4]
```diff
@@ -62,3 +68,122 @@ export const isExchangeSupported = (
...
+export const getPriceSource = async (
+  exchange: any,
+  pair: string,
+  priceSourceType: PriceSourceType,
+) => {
+  const orderBook = await exchange.fetchOrderBook(pair);
+  const ticker =
+    priceSourceType === PriceSourceType.LAST_PRICE
+      ? await exchange.fetchTicker(pair)
+      : null;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/utils/trading-strategy.utils.ts:91]: Why is ticker `null` if priceSourceType != `PriceSourceType.LAST_PRICE`

File: packages/tse-backend/src/common/utils/trading-strategy.utils.ts [hunk 5]
```diff
@@ -62,3 +68,122 @@ export const isExchangeSupported = (
...
+      return initialOrderAmount + (layer - 1) * amountChangePerLayer;
+    } else if (amountChangeType === AmountChangeType.PERCENTAGE) {
+      let adjustedAmount = initialOrderAmount;
+      for (let i = 1; i < layer; i++) {
+        adjustedAmount += adjustedAmount * (amountChangePerLayer / 100);
+      }
+      return adjustedAmount;
+    }
+  }
+  return initialOrderAmount;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/utils/trading-strategy.utils.ts:118]: Simplify that, we don't do nested conditions

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  async createMarketMakingStrategy(@Body() dto: MarketMakingStrategyDto) {
+    const command = this.mapper.map(
+      dto,
+      MarketMakingStrategyDto,
+      MarketMakingStrategyCommand,
+    );
+    return await this.service.create(command);
+  }
+
+  @Post('/pause-market-making')
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts:31]: Should that be `POST`?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  async pauseMarketMakingStrategy(@Query() dto: MarketMakingStrategyActionDto) {
+    const command = this.mapper.map(
+      dto,
+      MarketMakingStrategyActionDto,
+      MarketMakingStrategyActionCommand,
+    );
+    return await this.service.pause(command);
+  }
+
+  @Post('/stop-market-making')
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts:41]: Should that be POST?

Comment (piotrswierzy) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts:41]: What's the diffrence between: Pause & Stop?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  async stopMarketMakingStrategy(@Query() dto: MarketMakingStrategyActionDto) {
+    const command = this.mapper.map(
+      dto,
+      MarketMakingStrategyActionDto,
+      MarketMakingStrategyActionCommand,
+    );
+    return await this.service.stop(command);
+  }
+
+  @Post('/delete-market-making')
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts:51]: Should that be POST?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+      floorPrice: command.floorPrice,
+      status: StrategyInstanceStatus.RUNNING,
+    });
+  }
+
+  async pause(command: MarketMakingStrategyActionCommand): Promise<void> {
+    const strategyEntity: MarketMakingStrategyData =
+      await this.marketMakingService.findLatestStrategyByUserId(command.userId);
+    if (!strategyEntity) {
+      throw new BadRequestException('MarketMaking strategy not found');
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:62]: Shouldn't that be `404`?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    this.logger.debug('Paused market making strategy');
+  }
+
+  async stop(command: MarketMakingStrategyActionCommand): Promise<void> {
+    const strategyEntity: MarketMakingStrategyData =
+      await this.marketMakingService.findLatestStrategyByUserId(command.userId);
+    if (!strategyEntity) {
+      throw new BadRequestException('MarketMaking strategy not found');
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:81]: Shouldn't that be 404?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+  async delete(command: MarketMakingStrategyActionCommand) {
+    const strategyEntity: MarketMakingStrategyData =
+      await this.marketMakingService.findLatestStrategyByUserId(command.userId);
+    if (!strategyEntity) {
+      throw new BadRequestException('MarketMaking strategy not found');
+    }
+    if (
+      strategyEntity &&
+      strategyEntity.status !== StrategyInstanceStatus.STOPPED
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:109]: Why can't you delete a stopped strategy?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+  }
+
+  async start(strategies: MarketMakingStrategyData[]): Promise<void> {
+    this.logger.debug('Starting market making strategy');
+
+    for (const strategy of strategies) {
+      if (!this.strategies.get(strategy.id)) {
+        const intervalId = setInterval(async () => {
+          await this.evaluateMarketMaking(strategy);
+        }, strategy.checkIntervalSeconds * 1000);
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:137]: `1000` - magic number, should be from .env

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+      orderAmount,
+      numberOfLayers,
+      priceSourceType,
+      amountChangePerLayer,
+      amountChangeType,
+      ceilingPrice,
+      floorPrice,
+    } = command;
+
+    await this.cancelActiveOrders();
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:170]: Why is this here?

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+      amountChangeType,
+      amountChangePerLayer,
+      bidSpread,
+      askSpread,
+      priceSource,
+      ceilingPrice,
+      floorPrice,
+    );
+
+    for (const detail of orderDetails) {
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:188]: This for should be simplified, maybe extract parts to seperate functions

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/spec/market-making.strategy.spec.ts [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+        priceSourceType: PriceSourceType.MID_PRICE,
+        amountChangePerLayer: 0.05,
+        amountChangeType: AmountChangeType.PERCENTAGE,
+        ceilingPrice: 2000,
+        floorPrice: 1500,
+        createdAt: new Date(),
+        updatedAt: new Date(),
+        status: StrategyInstanceStatus.STOPPED,
+      };
+
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/trading-strategy/strategies/market-making/spec/market-making.strategy.spec.ts:238]: Extract to fixtures

**Review by piotrswierzy — CHANGES_REQUESTED** [review]

File: packages/tse-backend/src/common/entities/market-making.entity.ts (hunk 1 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/entities/market-making.entity.ts]: Fixed: ca125584

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts (hunk 11 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts]: Fixed: ca125584

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/spec/market-making.strategy.spec.ts (hunk 15 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/spec/market-making.strategy.spec.ts]: Fixed: ca125584

File: packages/tse-backend/src/common/entities/market-making.entity.ts (hunk 2 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/entities/market-making.entity.ts:47]: Layers represents market-making orders placed at different price levels, with varying order amounts and spreads, around the current market price. This approach aims to increase liquidity and profit through frequent, small trades at multiple price levels.

File: packages/tse-backend/src/common/utils/trading-strategy.utils.ts (hunk 4 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/trading-strategy.utils.ts:91]: The ticker is set to null for price sources other than LAST_PRICE because only the order book is needed for those, so there's no need to fetch the ticker:
```
  [PriceSourceType.BEST_BID]: (orderBook) => orderBook.bids[0][0],
  [PriceSourceType.LAST_PRICE]: (_, ticker) => ticker.last,
```

File: packages/tse-backend/src/common/utils/trading-strategy.utils.ts (hunk 3 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/trading-strategy.utils.ts:80]: Fixed: 1d377968

File: packages/tse-backend/src/common/utils/trading-strategy.utils.ts (hunk 5 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/trading-strategy.utils.ts]: Fixed: 96d2df91

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts (hunk 7 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts:41]: The Pause method stops the strategy execution (clear the interval),
The Stop method stops the strategy execution and cancels active (unfilled) orders

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts (hunk 6 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts]: Fixed: 04b08ce7

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts (hunk 7 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts]: Fixed: 04b08ce7

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts (hunk 8 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.controller.ts]: Fixed: 04b08ce7

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts (hunk 9 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts]: Fixed: b43902bf

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts (hunk 12 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts]: Fixed: 254babb4

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts (hunk 13 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:164]: Cancelling unfilled orders allows the strategy to better adjust new orders to the current market conditions in the next iteration

File: packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts (hunk 14 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/trading-strategy/strategies/market-making/market-making.strategy.ts:183]: Fixed: 131295c6
